From 8bdd534e665fcdea94d9b3ff903e6a001b0ffa33 Mon Sep 17 00:00:00 2001
From: Stanislav Kinsbursky <skinsbursky@parallels.com>
Date: Mon, 14 Feb 2011 17:44:41 +0300
Subject: [PATCH 1/2] sunrpc: ve semaphore deadlock fixed

Backported from rhel6
We need to release semaphore before return from xs_tcp_setup_socket

http://bugzilla.openvz.org/show_bug.cgi?id=1626

Signed-off-by: Stanislav Kinsbursky <skinsbursky@parallels.com>
Signed-off-by: Pavel Emelyanov <xemul@openvz.org>
---
 net/sunrpc/xprtsock.c |    2 ++
 1 files changed, 2 insertions(+), 0 deletions(-)

diff --git a/net/sunrpc/xprtsock.c b/net/sunrpc/xprtsock.c
index 8907690..2548df3 100644
--- a/net/sunrpc/xprtsock.c
+++ b/net/sunrpc/xprtsock.c
@@ -1971,6 +1971,8 @@ static void xs_tcp_setup_socket(struct rpc_xprt *xprt,
 	case -EINPROGRESS:
 	case -EALREADY:
 		xprt_clear_connecting(xprt);
+		up_read(&xprt->owner_env->op_sem);
+		(void)set_exec_env(ve);
 		return;
 	case -EINVAL:
 		/* Happens, for instance, if the user specified a link
-- 
1.7.4.1

